src/main/java/bio/overture/ego/service/RefreshContextService.java

```diff
@@ -0,0 +1,169 @@
+  }
+
+  private void checkUniqueByUser(User user) {
+    checkUnique(
+        !getRefreshTokenByUser(user).isPresent(),
+        String.format("A refresh token already exists for %s", user.getId()));
+  }
+
+  public void disassociateUserAndDelete(String userToken) {
+    val user = tokenService.getTokenUserInfo(userToken);
+
+    val refreshTokenOpt = this.getRefreshTokenByUser(user);
+    if (refreshTokenOpt.isPresent()) {
+      log.debug("Refresh token exists, deleting...");
+      user.setRefreshToken(null);
+      val existingRefreshToken = refreshTokenOpt.get();
+      refreshTokenRepository.delete(existingRefreshToken);
+    }
+  }
+
+  public RefreshToken createRefreshToken(@NonNull String bearerToken) {
+    val user = tokenService.getTokenUserInfo(bearerToken);
+    this.checkUniqueByUser(user);
+
+    val jti = tokenService.getTokenClaims(bearerToken).getId();
+    val refreshToken = createTransientToken(UUID.fromString(jti));
+    refreshToken.associateWithUser(user);
+    return refreshTokenRepository.save(refreshToken);
+  }
+
+  public RefreshContext createRefreshContext(String refreshTokenId, String bearerToken) {
+    val refreshTokenOpt = refreshTokenRepository.findById(UUID.fromString(refreshTokenId));
+    val tokenClaims = tokenService.getTokenClaims(bearerToken);
+    val user = tokenService.getTokenUserInfo(bearerToken);
+
+    checkNotFound(
+        !isNull(refreshTokenOpt) && refreshTokenOpt.isPresent(),
+        "RefreshToken %s is not found.",
+        refreshTokenId);
+
+    val refreshToken = refreshTokenOpt.get();
+    return RefreshContext.builder()
+        .refreshToken(refreshToken)
+        .tokenClaims(tokenClaims)
+        .user(user)
+        .build();
+  }
+
+  public RefreshContext createInitialRefreshContext(@NonNull String bearerToken) {
+    disassociateUserAndDelete(bearerToken);
+
+    if (tokenService.getTokenUserInfo(bearerToken).getStatus() != APPROVED) {
+      throw new ForbiddenException("User does not have approved status, rejecting.");
+    }
+
+    val newRefreshToken = createRefreshToken(bearerToken);
+    return createRefreshContext(newRefreshToken.getId().toString(), bearerToken);
+  }
+
+  public String validateAndReturnNewUserToken(String refreshId, String bearerToken) {
+    val incomingRefreshContext = createRefreshContext(refreshId, bearerToken);
+    disassociateUserAndDelete(bearerToken);
+
+    incomingRefreshContext.validate();
+
+    val newUserToken = tokenService.generateUserToken(incomingRefreshContext.getUser());
+    createRefreshToken(newUserToken);
+    return newUserToken;
+  }
+
+  private Cookie createCookie(String cookieName, String cookieValue, Integer maxAge) {
+    Cookie cookie = new Cookie(cookieName, cookieValue);
+    // where to access the accepted domain?
+    cookie.setDomain("localhost");
+    // disable setSecure while testing locally in browser, or will not show in cookies
+    //    cookie.setSecure(true);
```

Comment:
lets make this a configuration value in the application.yml

That way in the dev/test profile we can have this set to false.